File: Leetcode/js/21_mergeTwoSortedLists.js
/*
Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes of the first two lists.
*/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**a
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    if(l1 === null && l2 !== null) {
        return l2;
    } else if(l1 !== null && l2 === null) {
        return l1;
    } else if(l1 === null && l2 === null) {
        return null;
    }
    
    var head = null;
    if(l1.val <= l2.val){
        head = l1;
        l1 = l1.next;
    } else {
        head = l2;
        l2 = l2.next;
    }
    
    var tail = head;
    
    while(l1 !== null || l2 !== null) {
        if(l1 === null){
            tail.next = l2;
            tail = l2;
            l2 = l2.next;
        } else if(l2 === null) {
            tail.next = l1;
            tail = l1;
            l1 = l1.next;
        } else if(l1.val <= l2.val){
            tail.next = l1;
            tail = l1;
            l1 = l1.next;
        } else {
            tail.next = l2;
            tail = l2;
            l2 = l2.next;
        }
    }
    
    return head;
};

File: Leetcode/java/21_mergeTwoSortedLists.java
/*
Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes of the first two lists.
*/

class Main {
    public static void main(String[] args) {
        System.out.println("Starting...");

        Solution solution = new Solution();

        ListNode input1 = new ListNode(1);
        input.next = new ListNode(3);
        ListNode input2 = new ListNode(2);
        
        ListNode result = solution.mergeTwoLists(input1, input2);

        System.out.println("DONE!");
    }
}

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(-1);
        ListNode current = head;

        while(l1 != null && l2 != null) {
            if(l1.val < l2.val) {
                current.next = l1;
                l1 = l1.next;
            } else {
                current.next = l2;
                l2 = l2.next;
            }
            current = current.next;
        }

        while(l1 != null) {
            current.next = l1;
            l1 = l1.next;
            current = current.next;
        }

        while(l2 != null) {
            current.next = l2;
            l2 = l2.next;
            current = current.next;
        }

        return head.next;
    }
}












